lib/datadog/tracing/distributed/b3_single.rb

```diff
@@ -0,0 +1,67 @@
+# frozen_string_literal: true
+# typed: true
+
+require_relative 'helpers'
+require_relative '../trace_digest'
+
+module Datadog
+  module Tracing
+    module Distributed
+      # B3 single header-style trace propagation.
+      #
+      # DEV: Format:
+      # DEV:   b3: {TraceId}-{SpanId}-{SamplingState}-{ParentSpanId}
+      # DEV: https://github.com/apache/incubator-zipkin-b3-propagation/tree/7c6e9f14d6627832bd80baa87ac7dabee7be23cf#single-header
+      # DEV: `{SamplingState}` and `{ParentSpanId`}` are optional
+      #
+      # @see https://github.com/openzipkin/b3-propagation#single-header
+      class B3Single
+        def initialize(
+          key: Ext::B3_HEADER_SINGLE,
+          fetcher: Fetcher
+        )
+          @key = key
+          @fetcher = fetcher
+        end
+
+        def inject!(digest, env)
+          return if digest.nil?
+
+          # DEV: We need these to be hex encoded
+          value = "#{digest.trace_id.to_s(16)}-#{digest.span_id.to_s(16)}"
+
+          if digest.trace_sampling_priority
+            sampling_priority = Helpers.clamp_sampling_priority(
+              digest.trace_sampling_priority
+            )
+            value += "-#{sampling_priority}"
+          end
+
+          env[@key] = value
+          env
+        end
+
+        def extract(env)
+          fetcher = @fetcher.new(env)
+          value = fetcher[@key]
+
+          return unless value
+
+          parts = value.split('-')
+          trace_id = Helpers.value_to_id(parts[0], 16) unless parts.empty?
+          span_id = Helpers.value_to_id(parts[1], 16) if parts.length > 1
+          sampling_priority = Helpers.value_to_number(parts[2]) if parts.length > 2
```

Comment:
I believe these should go through the `fetcher`, right? Or otherwise, should we just drop the `id` and `number` methods from the fetcher and have callers use the `Helpers` directly?

Reply:
This is debatable: the fetcher knows to get data from the provided input `data`. It's currently a `key/value` accessor adapter.

This code her is a subdivision of a `value`, returned from a `key/value` pair.
Because this subdivision has nothing to do with retrieving the value from the `key/value` `data` input, I'm inclined to leave this as is. But as I said, this is debatable.

The fetcher is also protocol-specific (HTTP, gRPC), not style-specific (B3, Datadog) today, so _currently_ I think it makes sense to leave this as is. But this classification of the fetcher I just mentioned is arbitrary so 🤷